scanner/providers/gitcoin/gitcoin_provider.go

```diff
@@ -0,0 +1,336 @@
+package gitcoin
+
+import (
+	"bufio"
+	"context"
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/http"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/vocdoni/census3/scanner/providers"
+	"go.vocdoni.io/dvote/log"
+)
+
+const (
+	dateLayout      = "2006-01-02T15:04:05.999Z"
+	hexAddress      = "0x000000000000000000000000000000000000006C"
+	defaultCooldown = time.Hour * 6
+)
+
+type gitcoinScoreResult struct {
+	Passport struct {
+		Address string `json:"address"`
+	} `json:"passport"`
+	Score       string         `json:"score"`
+	StampScores map[string]any `json:"stamp_scores"`
+	Date        string         `json:"last_score_timestamp"`
+}
+
+type GitcoinPassport struct {
+	// public endpoint to download the json
+	apiEndpoint string
+	cooldown    time.Duration
+	// internal vars to manage the download
+	ctx         context.Context
+	cancel      context.CancelFunc
+	downloading *atomic.Bool
+	updated     *atomic.Bool
+	// internal vars to manage the balances
+	newBalances        map[common.Address]*big.Int
+	newBalancesMtx     sync.RWMutex
+	currentBalances    map[common.Address]*big.Int
+	currentBalancesMtx sync.RWMutex
+	lastUpdate         atomic.Value
+}
+
+type GitcoinPassportConf struct {
+	APIEndpoint string
+	Cooldown    time.Duration
+}
+
+// Init initializes the Gitcoin Passport provider with the given config. If the
+// config is not of type GitcoinPassportConf, or the API endpoint is missing, it
+// returns an error. If the cooldown is not set, it defaults to 6 hours.
+func (g *GitcoinPassport) Init(iconf any) error {
+	conf, ok := iconf.(GitcoinPassportConf)
+	if !ok {
+		return fmt.Errorf("invalid config type")
+	}
+	if conf.APIEndpoint == "" {
+		return fmt.Errorf("missing API endpoint")
+	}
+	if conf.Cooldown == 0 {
+		conf.Cooldown = defaultCooldown
+	}
+	g.apiEndpoint = conf.APIEndpoint
+	g.cooldown = conf.Cooldown
+	// init download variables
+	g.ctx, g.cancel = context.WithCancel(context.Background())
+	g.downloading = new(atomic.Bool)
+	g.updated = new(atomic.Bool)
+	g.downloading.Store(false)
+	g.updated.Store(false)
+	// init balances variables
+	g.currentBalances = make(map[common.Address]*big.Int)
+	g.currentBalancesMtx = sync.RWMutex{}
+	g.newBalances = make(map[common.Address]*big.Int)
+	g.newBalancesMtx = sync.RWMutex{}
+	g.lastUpdate.Store(time.Time{})
```

Comment:
no need to initialize and empty time, if you try to load later will return nil if not set, and store can be done lately with something like `g.lastUpdate.Store(time.Now())`